File: apps/node-local-agent/src/framework.ts
/**
 * REAL framework — no stubs. Mock backend included.
 */

export interface HarnessConfig {
  modelPath: string;
  workspaceDir: string;
  maxRetries: number;
  strictGrammar: boolean;
}

export interface Context {
  sessionId: string;
  memory: string[];
  toolResults: Map<string, string>;
}

export interface State {
  checkpoint: string;
  attempts: number;
}

export class HarnessError extends Error {
  constructor(public type: 'MockNotReady' | 'VerificationFailed' | 'RollbackError', message?: string) {
    super(message || type);
    this.name = 'HarnessError';
  }
}

export interface DomainHarness {
  name(): string;
  init(cfg: HarnessConfig): void;
  run(input: string, ctx: Context): Promise<string> | string;
  verify(output: string): boolean;
  rollback(state: State): State;
}

export class MockBackend {
  generate(prompt: string): string {
    return `MOCK_INFERENCE_RESULT: ${prompt.trim()}`;
  }
}

File: scripts/tmux_roco/framework.py
"""
Framework — Python mirror of crates/harness/src/framework.rs

DomainHarness trait, HarnessConfig, Context, State, HarnessError, ExecutionLoop
"""
from __future__ import annotations
from dataclasses import dataclass, field
from typing import Dict, List, Optional, Tuple
from abc import ABC, abstractmethod
import time

from .mock_backend import MockBackend


@dataclass
class HarnessConfig:
    model_path: str = "rwkv_mock"
    workspace_dir: str = "/tmp/mock"
    max_retries: int = 3
    strict_grammar: bool = True


@dataclass
class Context:
    session_id: str = "default"
    memory: List[str] = field(default_factory=list)
    tool_results: Dict[str, str] = field(default_factory=dict)


@dataclass
class State:
    checkpoint: str = ""
    attempts: int = 0

    def clone(self):
        return State(checkpoint=self.checkpoint, attempts=self.attempts)


class HarnessError(Exception):
    MockNotReady = "mock inference not connected"
    VerificationFailed = "verify failed"
    RollbackError = "rollback error"


class DomainHarness(ABC):
    """Abstract base, mirrors Rust trait."""

    @abstractmethod
    def name(self) -> str:
        ...

    def init(self, cfg: HarnessConfig) -> None:
        pass

    @abstractmethod
    def run(self, input_str: str, ctx: Context) -> str:
        ...

    @abstractmethod
    def verify(self, output: str) -> bool:
        ...

    @abstractmethod
    def rollback(self, state: State) -> State:
        ...

    def detailed_run(self, input_str: str) -> str:
        return f"[MOCK_DETAILED_{input_str}] input_length={len(input_str)} output_generated"


@dataclass
class LoopResult:
    output: str
    success: bool
    attempts: int
    rollback_count: int
    final_state: State
    history: List[State] = field(default_factory=list)


class ExecutionLoop:
    """Mirrors crates/harness/src/loop/mod.rs"""

    def __init__(self, max_attempts: int = 3):
        self.max_attempts = max_attempts

    def execute(self, agent: DomainHarness, input_str: str, ctx: Context) -> LoopResult:
        state = State()
        history: List[State] = []
        output = ""
        success = False

        for attempt in range(self.max_attempts):
            try:
                r = agent.run(input_str, ctx)
                output = r
                if agent.verify(output):
                    success = True
                    state.attempts = attempt + 1
                    state.checkpoint = f"check_{attempt}"
                    break
                else:
                    state = agent.rollback(state)
                    history.append(state.clone())
            except Exception:
                state = agent.rollback(state)
                history.append(state.clone())

        return LoopResult(
            output=output,
            success=success,
            attempts=state.attempts,
            rollback_count=len(history),
            final_state=state,
            history=history,
        )
